lib/wlib/memory/Memory.h

```diff
@@ -34,59 +40,136 @@ class MemoryInitDestroy {
     static int m_srefCount;
 };
 
+/**
+ * This code should never be called and/or touched. It is used internally
+ */
 static MemoryInitDestroy g_smemoryInitDestroy;
 
-extern "C" {
-
+/**
+ * NEVER EVER USE THIS OR I WILL COME AND CREATE A NEW VERSION OF YOU
+ * Use the @code malloc @endcode function
+ */
+void *memory_alloc(uint32_t size);
 
 /**
- * This function should and must be called exactly one time before application starts.
- * On C++ client code this is done automatically using MemoryInitDestroy
+ * NEVER EVER USE THIS OR I WILL COME AND REALLOCATE YOU
+ * Use the @code realloc @endcode function
  */
-void memory_init();
+void *memory_realloc(void *ptr, uint32_t size);
 
 /**
- * This function should and must be called exactly one time when the application exits. This should never
- * be called manually when using C++
+ * NEVER EVER USE THIS OR I WILL COME AND FREE YOU FROM LIFE
+ * Use the @code free @endcode function
  */
-void memory_destroy();
+void memory_free(void *ptr);
 
 /**
  * This allocates memory of the size provided. Memory allocated could be greater than what has been
- * asked in order to accommodate fixed memory allocations
+ * asked in order to accommodate fixed memory allocations. If there is not a sufficient memory
+ * available, it returns a @code nullptr @endcode
+ *
+ * @pre this can also act a @code new @endcode keyword to create objects so there is no need to use
+ * new keyword to create an object
  *
+ * @tparam type pointer type
  * @param size size of the block to allocate
  * @return address to memory allocated
  */
-void *memory_alloc(size_t size);
+template<typename type>
+type *malloc(uint32_t size = 1) {
+    void *memory = memory_alloc(sizeof(type) * size);
+    return new(memory) type;
+}
+
+/**
+ * This reallocates the memory to accommodate the new size provided. The memory address has to be
+ * allocation from Memory otherwise results are undefined. If there is not a sufficient memory
+ * available, it returns a @code nullptr @endcode
+ *
+ * @tparam type pointer type
+ * @param ptr address to memory to be reallocated
+ * @param size of new memory block
+ * @return address to new memory address
+ */
+template<typename type>
+type *realloc(type *ptr, uint32_t size = sizeof(type)) {

```

Comment:
`malloc` takes 1 parameter that is an integer multiple of `sizeof(Type)` but `realloc` takes as the size parameter the actual size in bytes. Should pick one and make it consistent for both, also make it really clear in the documentation.

Reply:
Oops just realized that I ma fix that now